<a href="https://colab.research.google.com/github/radwaelreedy/DEPI_Tech_Journey/blob/main/Assignment_KNN_VS_LogisticsRegression/Assignment_KNN_VS_LogisticsRegression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [ ]:
df = pd.read_csv("diabetes.csv")
df.head()

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
0,6,148,72,35,0,33.6,0.627,50,1
1,1,85,66,29,0,26.6,0.351,31,0
2,8,183,64,0,0,23.3,0.672,32,1
3,1,89,66,23,94,28.1,0.167,21,0
4,0,137,40,35,168,43.1,2.288,33,1


In [ ]:
df.shape

(768, 9)

In [ ]:
df.isin(['0']).sum()

,0
Pregnancies,0
Glucose,0
BloodPressure,0
SkinThickness,0
Insulin,0
BMI,0
DiabetesPedigreeFunction,0
Age,0
Outcome,0


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [ ]:
df.describe()

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
count,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000
mean,3.845052,120.894531,69.105469,20.536458,79.799479,31.992578,0.471876,33.240885,0.348958
std,3.369578,31.972618,19.355807,15.952218,115.244002,7.884160,0.331329,11.760232,0.476951
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.078000,21.000000,0.000000
25%,1.000000,99.000000,62.000000,0.000000,0.000000,27.300000,0.243750,24.000000,0.000000
50%,3.000000,117.000000,72.000000,23.000000,30.500000,32.000000,0.372500,29.000000,0.000000
75%,6.000000,140.250000,80.000000,32.000000,127.250000,36.600000,0.626250,41.000000,1.000000
max,17.000000,199.000000,122.000000,99.000000,846.000000,67.100000,2.420000,81.000000,1.000000


In [ ]:
info = pd.DataFrame(df.isnull().sum(), columns =["IsNull"])
info.insert(1,"isNa",df.isna().sum(),True)
info.insert(2,"Duplicate",df.duplicated().sum(),True)
info.insert(2,"Unique",df.nunique(),True)
info.T

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
IsNull,0,0,0,0,0,0,0,0,0
isNa,0,0,0,0,0,0,0,0,0
Unique,17,136,47,51,186,248,517,52,2
Duplicate,0,0,0,0,0,0,0,0,0


In [ ]:
cols_with_zero = ["Glucose", "BloodPressure", "SkinThickness", "Insulin", "BMI"]

for col in cols_with_zero:
    df[col] = df[col].replace(0, df[col].mean())

In [ ]:
df.info()
df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    float64
 2   BloodPressure             768 non-null    float64
 3   SkinThickness             768 non-null    float64
 4   Insulin                   768 non-null    float64
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(6), int64(3)
memory usage: 54.1 KB


,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
count,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000
mean,3.845052,121.681605,72.254807,26.606479,118.660163,32.450805,0.471876,33.240885,0.348958
std,3.369578,30.436016,12.115932,9.631241,93.080358,6.875374,0.331329,11.760232,0.476951
min,0.000000,44.000000,24.000000,7.000000,14.000000,18.200000,0.078000,21.000000,0.000000
25%,1.000000,99.750000,64.000000,20.536458,79.799479,27.500000,0.243750,24.000000,0.000000
50%,3.000000,117.000000,72.000000,23.000000,79.799479,32.000000,0.372500,29.000000,0.000000
75%,6.000000,140.250000,80.000000,32.000000,127.250000,36.600000,0.626250,41.000000,1.000000
max,17.000000,199.000000,122.000000,99.000000,846.000000,67.100000,2.420000,81.000000,1.000000


In [ ]:
X = df.drop("Outcome", axis=1)
y = df["Outcome"]

Why is replacing zero values better than dropping rows in this dataset?
No significant loss of data

In [ ]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

Why is scaling critical for KNN?
KNN works by calculating distances (usually Euclidean) between points to find the "nearest neighbors." If features are on different scales, those with larger ranges will dominate the distance calculation.

Why does Logistic Regression also benefit from scaling?
Scaling is highly beneficial—and often necessary—for two primary reasons: optimization efficiency and fair regularization

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X_scaled, y, test_size=0.25, random_state=42
)

Why do we evaluate models on test data instead of training data?
The Fundamental Reason: Generalization Assessment

Training data evaluation tells us how well the model learned the specific examples it saw. Test data evaluation tells us how well it generalizes to the real world.

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train, y_train)

KNeighborsClassifier()

In [ ]:
from sklearn.linear_model import LogisticRegression

lr = LogisticRegression(max_iter=1000)
lr.fit(X_train, y_train)

LogisticRegression(max_iter=1000)

In [ ]:
y_pred_knn = knn.predict(X_test)
y_pred_lr = lr.predict(X_test)

Identify TP, FP, FN, TN for each model

Positive (P) → Patient has diabetes

Negative (N) → Patient does NOT have diabetes

True Positive (TP)

Model predicts diabetes

Patient actually has diabetes

False Positive (FP)

Model predicts diabetes

Patient does NOT have diabetes

False Negative (FN)

Model predicts no diabetes

Patient actually has diabetes

True Negative (TN)

Model predicts no diabetes

Patient does NOT have diabetes

Which error is more dangerous in diabetes detection?

Patient has diabetes but is told they are healthy

No treatment → disease progresses silently

In [ ]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, y_pred_knn), confusion_matrix(y_test, y_pred_lr)

(array([[96, 27],
        [24, 45]]),
 array([[99, 24],
        [27, 42]]))

In [ ]:
from sklearn.metrics import accuracy_score

print("KNN Accuracy:", accuracy_score(y_test, y_pred_knn))
print("Logistic Regression Accuracy:", accuracy_score(y_test, y_pred_lr))

KNN Accuracy: 0.734375
Logistic Regression Accuracy: 0.734375


What does precision measure?

Precision=TP/TP+FP

What does recall measure?

Recall=TP/TP+FN

Why is F1-score often better than accuracy?

Accuracy can be misleading for imbalanced data

F1=2×(Precision×Recall/Precision+Recall)

Balances precision AND recall

In [ ]:
from sklearn.metrics import classification_report

print("KNN")
print(classification_report(y_test, y_pred_knn))

print("Logistic Regression")
print(classification_report(y_test, y_pred_lr))

KNN
              precision    recall  f1-score   support

           0       0.80      0.78      0.79       123
           1       0.62      0.65      0.64        69

    accuracy                           0.73       192
   macro avg       0.71      0.72      0.71       192
weighted avg       0.74      0.73      0.74       192

Logistic Regression
              precision    recall  f1-score   support

           0       0.79      0.80      0.80       123
           1       0.64      0.61      0.62        69

    accuracy                           0.73       192
   macro avg       0.71      0.71      0.71       192
weighted avg       0.73      0.73      0.73       192



1. Which model performed better overall?

KNN performed slightly better. While both models have the exact same accuracy (0.73)

2. Which model is easier to explain?

Logistic Regression is much easier to explain

3. Which model would struggle more with large datasets?

KNN would struggle significantly more.

4. If missing a diabetes case is very dangerous, which metric matters most?

Recall